x-pack/legacy/plugins/reporting/server/lib/export_types_registry.ts

```diff
@@ -4,40 +4,108 @@
+    if (!isString(item.id)) {
+      throw new Error(`'item' must have a String 'id' property `);
+    }
+
+    if (this._map.has(item.id)) {
+      throw new Error(`'item' with id ${item.id} has already been registered`);
+    }
+
+    this._map.set(item.id, item);
+  }
+
+  getAll() {
+    return Array.from(this._map.values());
+  }
+
+  getSize() {
+    return this._map.size;
+  }
+
+  getById<JobParamsType, CreateJobFnType, JobPayloadType, ExecuteJobFnType>(
+    id: string
+  ): ExportTypeDefinition<JobParamsType, CreateJobFnType, JobPayloadType, ExecuteJobFnType> {
+    if (!this._map.has(id)) {
+      throw new Error(`Unknown id ${id}`);
+    }
+
+    return this._map.get(id) as ExportTypeDefinition<
+      JobParamsType,
+      CreateJobFnType,
+      JobPayloadType,
+      ExecuteJobFnType
+    >;
+  }
+
+  get<JobParamsType, CreateJobFnType, JobPayloadType, ExecuteJobFnType>(
+    findType: GetCallbackFn<JobParamsType, CreateJobFnType, JobPayloadType, ExecuteJobFnType>
+  ): ExportTypeDefinition<JobParamsType, CreateJobFnType, JobPayloadType, ExecuteJobFnType> {
+    let result;
+    for (const value of this._map.values()) {
+      if (!findType(value)) {
+        continue; // try next value
       }
+      const foundResult: ExportTypeDefinition<
+        JobParamsType,
+        CreateJobFnType,
+        JobPayloadType,
+        ExecuteJobFnType
+      > = value;
 
-      resolve(files);
-    });
-  });
-}
+      if (result) {
+        throw new Error('Found multiple items matching predicate.');
+      }
+
+      result = foundResult;
+    }
 
-const pattern = pathResolve(__dirname, '../../export_types/*/server/index.[jt]s');
-async function exportTypesRegistryFn(server: ServerFacade) {
-  const logger = LevelLogger.createForServer(server, [PLUGIN_ID, 'exportTypes']);
-  const exportTypesRegistry = new ExportTypesRegistry();
-  const files: string[] = (await scan(pattern)) as string[];
+    if (!result) {
+      throw new Error('Found no items matching predicate');
+    }
+
+    return result;
+  }
+}
 
-  files.forEach(file => {
-    logger.debug(`Found exportType at ${file}`);
+function getExportTypesRegistryFn(): ExportTypesRegistry {
+  const registry = new ExportTypesRegistry();
 
-    const { register } = require(file); // eslint-disable-line @typescript-eslint/no-var-requires
-    register(exportTypesRegistry);
+  /* this replaces the previously async method of registering export types,
+   * where this would run a directory scan and types would be registered via
+   * discovery */
+  const getTypeFns: Array<() => ExportTypeDefinition<any, any, any, any>> = [
+    getTypeCsv,
+    getTypeCsvFromSavedObject,
+    getTypePng,
+    getTypePrintablePdf,
+  ];
+  getTypeFns.forEach(getType => {
+    registry.register(getType());
   });
-  return exportTypesRegistry;
+  return registry;
 }
 
-export const exportTypesRegistryFactory = oncePerServer(exportTypesRegistryFn);
+// FIXME: is this the best way to return a singleton?
```

Comment:
Maybe there is no need to memoize this, since `exportTypesRegistry` is consumed in 1 place

Reply:
Why not instanciate `ExportTypesRegistry` in the module and export the singleton then? Maybe a little harder for tests though.